pytorch_lightning/strategies/ddp.py

```diff
@@ -256,24 +255,47 @@ def _reinit_optimizers_with_post_localSGD(self, warmup_steps: int):
                 is_distributed_optimizer
                 or isinstance(optimizer, ZeroRedundancyOptimizer)
                 or (_FAIRSCALE_AVAILABLE and isinstance(optimizer, OSS))
+                or isinstance(optimizer, PostLocalSGDOptimizer)
             ):
                 raise ValueError(
-                    f"Cannot wrap a distributed optimizer of type {optimizer.__name__} by PostLocalSGDOptimizer."
+                    f"Currently model averaging cannot work with a distributed optimizer of type "
+                    f"{optimizer.__class__.__name__}."
                 )
 
-            if isinstance(optimizer, PostLocalSGDOptimizer):
-                continue
+        self._model_averager = torch.distributed.algorithms.model_averaging.averagers.PeriodicModelAverager(
+            period=self._model_averaging_period, warmup_steps=self._ddp_comm_state.start_localSGD_iter
+        )
 
-            optim_class = type(optimizer)
-            post_localSGD_optimizer = PostLocalSGDOptimizer(
-                params=optimizer.param_groups,
-                optimizer_class=optim_class,
-                averager=averager,
-                **optimizer.defaults,
-            )
-            optimizers[x] = post_localSGD_optimizer
-            del optimizer
-        self.optimizers = optimizers
+    def optimizer_step(
+        self,
+        optimizer: Optimizer,
+        opt_idx: int,
+        closure: Callable[[], Any],
+        model: Optional[Union["pl.LightningModule", Module]] = None,
+        **kwargs: Any,
+    ) -> Any:
+        """performs the actual optimizer step.
+
+        Args:
+            optimizer: the optimizer performing the step
+            opt_idx: index of the current optimizer
+            closure: closure calculating the loss value
+            model: reference to the model, optionally defining optimizer step related hooks
+            **kwargs: Any extra arguments to ``optimizer.step``
+        """
+        optimizer_output = super().optimizer_step(optimizer, opt_idx, closure, model, **kwargs)
+
+        if not _TORCH_GREATER_EQUAL_1_10 or self._model_averager is None:
+            return optimizer_output
+
+        for opt in self.optimizers:
```

Comment:
Is it correct? From the documentation, it seems we need to provide averager.average_parameters(model.parameters()) entirely. Won't this trigger several average step?

Reply:
@rohan-varma

Reply:
Sorry for the confusion here.

Actually the documentation is not accurate -- `averager.average_parameters(model.parameters())` is not always correct, because if a model has untrainable parameters, then we shouldn't average all the parameters, so the implementation here is more accurate.

Thus, `averager.average_parameters(model.parameters())` is not added to PyTorch Lightning, and the code here is a better replacement.

Reply:
Sounds good, thanks @wayi1 for the info!!

Reply:
Thanks :)